src/coinsselectionalgorithm.hpp

```diff
@@ -0,0 +1,280 @@
+#ifndef _COINS_SELECTION_ALGORITHM_H
+#define _COINS_SELECTION_ALGORITHM_H
+
+
+#include <string>
+#include <thread>
+#include <vector>
+#include "amount.h"
+
+
+//! Flag for profiling/debugging mode
+#define COINS_SELECTION_ALGORITHM_PROFILING 0
+//! This represents the number of intermediate change levels inside the interval [targetAmount + 0, targetAmount + maxChange]
+/*!
+  Low value -> higher quantity of selected utxos and higher change, high value -> lower quantity of selected utxos and lower change
+*/
+#define COINS_SELECTION_INTERMEDIATE_CHANGE_LEVELS 9
+
+
+//! Types of CCoinsSelectionAlgorithm
+enum class CoinsSelectionAlgorithmType {
+    UNDEFINED = 0,
+    SLIDING_WINDOW = 1,
+    BRANCH_AND_BOUND = 2
+};
+
+/* ---------- CCoinsSelectionAlgorithm ---------- */
+
+//! Abstract class for algorithm of coins selection
+/*!
+  This class provides common fields required by each implementation and utility methods
+*/
+class CCoinsSelectionAlgorithm
+{    
+protected:
+    // auxiliary
+    //! The temporary set of selected elements (true->selected, false->unselected)
+    bool* tempSelection;
```

Comment:
Your point makes totally sense to me, but there is a but...
Actually during the initial prototypal phase, all those pointers were `std::vector`, then I moved to performance tuning and I got a significant improvement moving to raw pointers: I am talking about 10%/15% less execution time.
Given that the selection algorithm are very performance critical I would keep the raw pointers implementation; for safety I'd ask to the reviewers a double check on this piece of code (@ptagl, @drgora, @a-petrini).

Reply:
Two considerations here :

- "**Premature optimization is the root of all evils**" (D. Knuth) : first we write safe and maintainable code then we work on it for optimization (if necessary)
- You can use `unique_ptr<boo[]>` at the very least

Reply:
I have pushed on remote a new temporary branch named *cso_profiling_test*, three commits are available:

- AS IS: current (as of 225568e0caa980a2ddce83c4844cc98ba154d655) implementation
- WITH VECTORS: all raw pointers data replaced by std:vector 
- WITH SMART POINTERS: all raw pointers data replaced by std:_unique_ptr

it's a very quick-and-dirt branch which allow an immediate test of algorithm selection routine at the very start of Zend node (actually aborting the start-up procedure), printing a rough report file with timing performance.

Running on my machine I confirm the performance degradation using both vectors or smart pointers in place of raw pointers. I know @drgora is playing with that branch and is reporting a degradation with vectors but not with smart pointers; let's wait for a feedback also by the other team members and then we will wrap up all the ideas and take a decision about which implementation to chose.

Reply:
Let's consider also the two algos, in this impl, do not run in parallel hence the overall performance is affected.

Reply:
Pushed new commit for moving from raw pointers to vectors. This allows safer memory management and better maintainability.

Regarding performances, I need to clarify and explain the root cause of the degradation I experienced using vectors during prototypal phase.
The focus is on members `CCoinsSelectionAlgorithmBase::tempSelection` and `CCoinsSelectionAlgorithmBase::optimalSelection`; when refactoring those members from being `bool*` to being `std::vector<bool>` I didn't consider that `std::vector<bool>` _"is a specialized version of [vector](https://cplusplus.com/vector), which is used for elements of type bool and optimizes for space"_. Indeed, each bool element is in some way compressed/optimized _"so that each value is stored in a single bit"_; hence it seems to me reasonable that accessing elements through the index operator `[]` has to deal with a small overhead, that would eventually result in that ~10% performance degradation. In the end I decided to define those two members as `std::vector<char>`, preferring to favor processing over memory optimization in this situation.
Thanks to @drgora for pointing this out and orienting me on the right way.
Additional reference: https://cplusplus.com/reference/vector/vector-bool/.

Reply:
That is generally correct with one hint : actually memory is not optimized nor compressed. Is simply how memory access works.  

The problem is a bool is actually 1 bit (0 or 1) but  `sizeof(bool)` is implementation defined generally to 1 byte as computers can't access memory in denomination units less than that.  Hence std::vector<bool> as well as std::bitset allocate a minimum of 1 byte.

Hance to retrieve the boolean in a position of the vector implies :
- a division to find the byte where the bit we want to examine is stored (BN)
- a modulo operation to find the bit within the byte (bN)
- a bitwise operation to determine if the bit is set or not : `BN & (1 << bN)`

This is the overhead. Also endianness plays a role as we need to identify is the bit in question is RtL or LtR